feat: SinglePrivateMutable

benesjan opened the pull request:
Fixes F-187

Just like `PrivateMutable` but allows only for 1 instance per contract instead of 1 instance per contract-user pair.

That 1 instance invariant is achieved by having a nullifier that has only the storage slot in it's preimage. This currently leaks that initialization happened - tackling this is a TODO.

benesjan commented:
* **#18824** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18824?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18824?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/noir-contracts/contracts/app/private_token_contract/src/main.nr:
```diff
@@ -5,12 +5,14 @@ pub contract PrivateToken {
     #[storage]
     struct Storage<Context> {
         balances: Owned<BalanceSet<Context>, Context>,
+        total_supply: GlobalPrivateMutable<FieldNote, Context>,
```
We didn't have any use case for the `GlobalPrivateMutable` in our codebase so I introduced this change.

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_mutable.nr:
```diff
@@ -0,0 +1,263 @@
+    {
+        let old_owner = self.context.msg_sender().unwrap();
+        let (prev_retrieved_note, note_hash_read) =
+            get_note(self.context, Option::some(old_owner), self.storage_slot);
```
As discussed, this hould be `none`.

benesjan replied:
Addressed

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/global_private_mutable.nr:
```diff
@@ -0,0 +1,263 @@
+
+/// GlobalPrivateMutable is a state variable that represents a contract-global private value that can be changed. Unlike
+/// [crate::state_vars::private_mutable::PrivateMutable] which is "owned" (requiring wrapping in an
+/// [crate::state_vars::owned::Owned] state variable), GlobalPrivateMutable is used directly in storage:
```
I'd expand on example use cases (acct contract, closed systems with no race condition issues like multisigs - mention race conditions). Expand on intended audience of the private state.

benesjan replied:
Put there 2 examples but I am not sure how this could be used in multisigs.

Anyway this feels quite twisted.

benesjan commented on noir-projects/noir-contracts/contracts/app/private_token_contract/src/main.nr:
```diff
@@ -1,46 +1,101 @@
-            MessageDelivery.CONSTRAINED_ONCHAIN,
-        );
+    fn mint(amount: u128, recipient: AztecAddress) {
+        let replacement_note_message = self.storage.admin.get_note();
```
`get_note` returning a note message feels weird (this is shared with `PrivateMutable` but surprisingly we don't ever seem to use the `get_note` method of `PrivateMutable` in contracts now).

I think the function name is bad as it doesn't communicate that the "replacement pattern" is happening in it.

nventuro replied:
>but surprisingly we don't ever seem to use the get_note method of PrivateMutable in contracts now).

I noticed this as well - we just use `replace` always it seems. Do you have any API suggestions?

benesjan replied:
[Discussed on slack](https://aztecprotocol.slack.com/archives/C07R3GTKJ49/p1765985821862809) and will potentially be addressed in a separate PR.

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_mutable.nr:
```diff
@@ -0,0 +1,261 @@
+/// not a possibility because there is no owner.
+///
+/// This issue is shared with contract initialization nullifiers, and is tracked by
+/// https://github.com/AztecProtocol/aztec-packages/issues/17128
```
I think it's quite likely that this state var will also prove desirable to be used in "closed set" settings where there could be a notion of shared keys which might make it possible to use the nsk of those shared keys in the initialization nullifier.

But would wait with tackling this as we don't have enough usecases yet so it all feels a bit abstract now.

(In the `SinglePrivateImmutable` I use the contract's `nsk`)

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_mutable.nr:
```diff
@@ -0,0 +1,261 @@
+/// anyone with the knowledge of the contract address (nullifiers are being hashed with the contract address by the
+/// kernel circuits before being inserted into the nullifier tree and the storage slot can be guessed/brute forced).
+/// This problem is generally avoided by including owner's nullifying secret in the nullifier preimage. Here this is
+/// not a possibility because there is no owner.
```
What if we used the contract's nsk?

benesjan replied:
Spent a bit of time thinking about this and I came to a conclusion that it's a good thing to do so I addressed it in https://github.com/AztecProtocol/aztec-packages/pull/18824/commits/126bdafbae4b8ab64fef84873b8b07e5cef9f729 and in https://github.com/AztecProtocol/aztec-packages/pull/18824/commits/d74e0d8f863609e07311db9fcc12bcca4eefc47d .

It feels like we don't support non-account-contracts-deployed-with-keys well enough because:
1. Deploying them looks ugly: https://github.com/AztecProtocol/aztec-packages/pull/18824/commits/d74e0d8f863609e07311db9fcc12bcca4eefc47d.
2. In the case of the `PrivateToken` we don't need to receive notes in the contract (we only need the nsk for initialization nullifier) but we need to register it as an account in PXE which adds a lot of cost due to tagging.
3. Now when interacting with the contract from a new wallet we will need to pass around not just the contract artifact + instance but also the keys.

I propose creating a `Closed Set Apps` project on linear and tracking the above issues there. Agree?

benesjan replied:
Nico doesn't want to look at this so went ahead and just created [it](https://linear.app/aztec-labs/project/closed-set-apps-f4eb9ec0d397/overview).

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_mutable/test.nr:
```diff
@@ -0,0 +1,406 @@
+    env.private_context_at(contract_address, |context| {
+        let state_var = in_private(context);
+
+        let _ = state_var.replace(
```
`replace` is a great name for this 🙌

mverzilli commented on noir-projects/noir-contracts/contracts/app/private_token_contract/src/main.nr:
```diff
@@ -1,46 +1,101 @@
+    fn constructor(initial_admin_balance: u128, admin: AztecAddress) {
+        // Admin itself is the owner of the underlying note because we want admin to be able to transfer the admin
+        // privileges to someone else which requires being able to nullify the note. We use constrained onchain message
+        // delivery because we don't know if the party deploying this contract is incentivized to deliver the note.
```
nice

mverzilli commented on noir-projects/noir-contracts/contracts/app/private_token_contract/src/main.nr:
```diff
@@ -1,46 +1,101 @@
+    }
+
+    // Transfers admin privileges to a `new_admin`.
+    #[external("private")]
+    fn transfer_admin(new_admin: AztecAddress) {
+        self
+            .storage
+            .admin
+            .replace(
+                |old| {
+                    assert(
+                        old.address == self.msg_sender().unwrap(),
+                        "Only admin can transfer admin privileges",
+                    );
+                    AddressNote { address: new_admin }
+                },
+                new_admin,
+            )
+            .deliver(MessageDelivery.CONSTRAINED_ONCHAIN);
     }
```
reads really nice

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_mutable.nr:
```diff
@@ -0,0 +1,269 @@
+///
+/// # Examples
+///
+/// ## Account contract with signing key rotation
+///
+/// An account contract's signing key can be modeled as a SinglePrivateMutable<PublicKeyNote>. The "current value" of
+/// this state variable holds the active signing key. When the account owner wishes to update the signing key, they
+/// invoke the `replace` function, providing a new public key note as input.
```
cool example

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_mutable.nr:
```diff
@@ -0,0 +1,269 @@
+    /// uninitialized SinglePrivateMutable, it will fail because there is no current note to replace. If you don't know
+    /// if the state variable has been initialized already, you can use `initialize_or_replace` to handle both cases.
+    ///
+    /// The nullification of the previous note ensures that it cannot be used again, maintaining the invariant that a
+    /// SinglePrivateMutable has exactly one current note.
```
nice

mverzilli reviewed (approved):
This ended up looking super neat, great job!

benesjan committed:
feat: SinglePrivateMutable

Fixes F-187

Just like `PrivateMutable` but allows only for 1 instance per contract instead of 1 instance per contract-user pair.

That 1 instance invariant is achieved by having a nullifier that has only the storage slot in it's preimage. This currently leaks that initialization happened - tackling this is a TODO.
